umh-core/pkg/communicator/actions/stateMocker.go

```diff
@@ -0,0 +1,486 @@
+	"github.com/united-manufacturing-hub/united-manufacturing-hub/umh-core/pkg/constants"
+	"github.com/united-manufacturing-hub/united-manufacturing-hub/umh-core/pkg/fsm"
+	"github.com/united-manufacturing-hub/united-manufacturing-hub/umh-core/pkg/fsm/dataflowcomponent"
+	"go.uber.org/zap"
+)
+
+// ErrAlreadyRunning is returned when trying to start a StateMocker that is already running
+var ErrAlreadyRunning = errors.New("state mocker already running")
+
+// ConfigManager interface defines the methods needed by StateMocker to get configuration data
+type ConfigManager interface {
+	// GetDataFlowConfig returns the DataFlow component configurations
+	GetDataFlowConfig() []config.DataFlowComponentConfig
+}
+
+// StateTransition represents a state transition for a component
+type StateTransition struct {
+	TickAt int
+	State  string
+}
+
+// the state mocker is used for unit testing to mock the state of the system
+// it runs in a separate goroutine and regularly updates the state of the system
+// it is passed a config manager that provides the config being used by the action unit tests
+// it then updates the state of the system according to the config just like the real system would do
+type StateMocker struct {
+	StateManager       *fsm.SnapshotManager
+	ConfigManager      ConfigManager
+	TickCounter        int
+	PendingTransitions map[string][]StateTransition // key is component ID
+	done               chan struct{}                // channel to signal shutdown of goroutine
+	running            atomic.Bool                  // flag to track if the mocker is running
+	mu                 *sync.RWMutex                // mutex to protect the state of the mocker
+	LastConfig         config.FullConfig            // last config is needed to detect config changes (events)
+	LastConfigSet      bool                         // flag to track if the last config has been set
+	IgnoreDfcUntilTick map[string]int               // map of component ID to tick to ignore
+}
+
+// NewStateMocker creates a new StateMocker
+func NewStateMocker(configManager ConfigManager) *StateMocker {
+	return &StateMocker{
+		ConfigManager:      configManager,
+		StateManager:       fsm.NewSnapshotManager(),
+		TickCounter:        0,
+		PendingTransitions: make(map[string][]StateTransition),
+		done:               make(chan struct{}),
+		mu:                 &sync.RWMutex{},
+		LastConfig:         config.FullConfig{},
+		LastConfigSet:      false,
+		IgnoreDfcUntilTick: make(map[string]int),
+	}
+}
+
+func (s *StateMocker) GetStateManager() *fsm.SnapshotManager {
+	return s.StateManager
+}
+
+// GetConfigManager returns the config manager that is being used by the state mocker
+func (s *StateMocker) GetConfigManager() ConfigManager {
+	return s.ConfigManager
+}
+
+// Tick advances the state mocker by one tick and updates the system state
+func (s *StateMocker) Tick() {
+	s.TickCounter++
+	s.UpdateDfcState()
+}
+
+// here, the actual state update logic is implemented
```

Comment:
add docstring comment what is happening here (can be made based on the inline comments)